core/bin/snapshots_creator/src/creator.rs

```diff
@@ -0,0 +1,335 @@
+        let sealed_l1_batch_number = sealed_l1_batch_number.context("No L1 batches in Postgres")?;
+        anyhow::ensure!(
+            sealed_l1_batch_number != L1BatchNumber(0),
+            "Cannot create snapshot when only the genesis L1 batch is present in Postgres"
+        );
+        let l1_batch_number = sealed_l1_batch_number - 1;
+
+        let latest_snapshot_l1_batch_number =
+            latest_snapshot.map(|snapshot| snapshot.l1_batch_number);
+        if latest_snapshot_l1_batch_number == Some(l1_batch_number) {
+            tracing::info!(
+                "Snapshot at expected L1 batch #{l1_batch_number} is already created; exiting"
+            );
+            return Ok(None);
+        }
+
+        let distinct_storage_logs_keys_count = conn
+            .snapshots_creator_dal()
+            .get_distinct_storage_logs_keys_count(l1_batch_number)
+            .await?;
+        let chunk_size = config.storage_logs_chunk_size;
+        // We force the minimum number of chunks to avoid situations where only one chunk is created in tests.
+        let chunk_count =
+            ceil_div(distinct_storage_logs_keys_count, chunk_size).max(min_chunk_count);
+
+        tracing::info!(
+            "Selected storage logs chunking for L1 batch {l1_batch_number}: \
+            {chunk_count} chunks of expected size {chunk_size}"
+        );
+        Ok(Some(SnapshotProgress::new(l1_batch_number, chunk_count)))
+    }
+
+    /// Returns `Ok(None)` if a snapshot should not be created / resumed.
+    async fn load_or_initialize_snapshot_progress(
+        &self,
+        config: &SnapshotsCreatorConfig,
+        min_chunk_count: u64,
+    ) -> anyhow::Result<Option<SnapshotProgress>> {
+        let mut master_conn = self
+            .master_pool
+            .access_storage_tagged("snapshots_creator")
+            .await?;
+        let latest_snapshot = master_conn
+            .snapshots_dal()
+            .get_newest_snapshot_metadata()
+            .await?;
+        drop(master_conn);
+
+        let pending_snapshot = latest_snapshot
+            .as_ref()
+            .filter(|snapshot| !snapshot.is_complete());
+        if let Some(snapshot) = pending_snapshot {
+            Ok(Some(SnapshotProgress::from_existing_snapshot(snapshot)))
+        } else {
+            Self::initialize_snapshot_progress(
+                config,
+                min_chunk_count,
+                latest_snapshot.as_ref(),
+                &mut self.connect_to_replica().await?,
+            )
+            .await
+        }
+    }
+
+    pub async fn run(
+        self,
+        config: SnapshotsCreatorConfig,
+        min_chunk_count: u64,
+    ) -> anyhow::Result<()> {
+        let latency = METRICS.snapshot_generation_duration.start();
+
+        let Some(progress) = self
+            .load_or_initialize_snapshot_progress(&config, min_chunk_count)
+            .await?
+        else {
+            return Ok(()); // No snapshot creation is necessary
```

Comment:
would it make sense to mention in the comment why it happens - that is, the snapshot for the current l1 batch is already genertaed?